http-clients/src/main/java/org/triplea/http/client/lobby/chat/messages/server/ChatServerMessageType.java

```diff
@@ -0,0 +1,36 @@
+package org.triplea.http.client.lobby.chat.messages.server;
+
+import com.google.common.base.Preconditions;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import lombok.AllArgsConstructor;
+import org.triplea.domain.data.PlayerName;
+import org.triplea.http.client.lobby.chat.ChatMessageListeners;
+import org.triplea.http.client.lobby.chat.ChatParticipant;
+import org.triplea.http.client.web.socket.messages.ServerMessageEnvelope;
+
+/** Chat message types that a server can send over websocket to client. */
+@AllArgsConstructor
+@SuppressWarnings("ImmutableEnumChecker")
+public enum ChatServerMessageType {
+  CHAT_EVENT(String.class, ChatMessageListeners::getChatEventListener),
+  CHAT_MESSAGE(ChatMessage.class, ChatMessageListeners::getChatMessageListener),
+  PLAYER_JOINED(ChatParticipant.class, ChatMessageListeners::getPlayerJoinedListener),
+  PLAYER_LEFT(PlayerName.class, ChatMessageListeners::getPlayerLeftListener),
+  PLAYER_LISTING(ChatterList.class, ChatMessageListeners::getConnectedListener),
+  PLAYER_SLAPPED(PlayerSlapped.class, ChatMessageListeners::getPlayerSlappedListener),
+  SERVER_ERROR(String.class, ChatMessageListeners::getServerErrorListener),
+  STATUS_CHANGED(StatusUpdate.class, ChatMessageListeners::getPlayerStatusListener);
+
+  private final Class<?> classType;
+  private final Function<ChatMessageListeners, Consumer<?>> extractListener;
+
+  @SuppressWarnings("unchecked")
+  public <T> void sendPayloadToListener(
+      final ServerMessageEnvelope serverMessageEnvelope,
+      final ChatMessageListeners chatMessageListeners) {
+    Preconditions.checkArgument(valueOf(serverMessageEnvelope.getMessageType()) == this);
+    final T payload = (T) serverMessageEnvelope.getPayload(classType);
+    ((Consumer<T>) extractListener.apply(chatMessageListeners)).accept(payload);
```

Comment:
Indeed, the generic issue is not ideal. As a concrete classtype, we'd lose the 'valueOf' functionality. That would be valuable, but I do think the enum syntax and having the 'valueOf' and not rolling-our-own is more valuable. 

Re: immutable enum warning, I'm thinking the check is flagging false positives. An enum with a function is immutable. The warning is checking that an enum is nothing but constants, which is not the same. I'd agree an enum should be immutable, though the check is not quite verifying that. The pattern to attach methods/functions to enum is pretty poweful, it was a major language decision that IMO was done for a reason. The anonymous inner class work-around is a less ideal approach to that even.

The method override variant I believe is essentially an anonymous class.  There is an effective java item that recommends to avoid the anonymous class in favor of lambda:

> Item 42: Prefer lambdas to anonymous classesHistorically, interfaces 

https://kea.nu/files/textbooks/new/Effective%20Java%20%282017%2C%20Addison-Wesley%29.pdf  (p193)